classes/local/envbarlib.php

```diff
@@ -329,7 +329,7 @@ public static function inject() {
                 'showtext' => get_string('prod', 'local_envbar'),
                 'colourtext' => get_config('local_envbar', 'prodtextcolour'),
                 'colourbg' => get_config('local_envbar', 'prodbgcolour'),
-                'matchpattern' => self::getprodwwwroot(),
+                'matchpattern' => rtrim(self::getprodwwwroot(), '/') . '/',
```

Comment:
Hi @peterspicer-catalyst 
Would be really good if you could move this code inside  getprodwwwroot() function and add some tests to actually understand what you are trying to fix here.

Reply:
Hi @dmitriim 
Moving it inside getprodwwwroot() actually complicates things because this explicitly needs different behaviours in different places.

I did try to explain it in the bug report, but let me try to paint a picture of the situation.

So, we have a Moodle whose production is something like https://mymoodle.example.com/ - and when I set up a test environment, it's going to get a URL similar to https://mymoodle.example.com.petersmachine.btn.catalyst-eu.net/ and it may well have a staging environment at https://mymoodle.example.com.client.catalyst-eu.net/

Currently, because the prod www root doesn't have a trailing slash, and is effectively matched as a prefix, all of these environments appear to be production as far as the envbar is concerned.

On the one hand, I don't want to alter the master prod www root in getprodwwwroot because then every place that uses that function (like link building) needs to remember to strip the trailing slash again, and I really only want to change it for the purposes of pattern matching so that prod and only prod gets the trailing slash to stop it being matched against non-prod environments that use the same domain as a prefix.

Updating just the pattern match for the prod environment means that nothing else is broken, or requires changes while making sure that production links based off prod www root are highlighted as expected.

Writing tests is also really difficult because the tests are pretty explicitly tied to the testing environment and simulating this in a testable way is something I couldn't figure out how to do since it really requires having two different URLs routable to the same Moodle, and being able to switch $CFG->wwwroot (or the Behat version if doing it that way) out mid flight which is technically possible in a Behat step definition but the step definition (or the feature file) is bound to that specific test environment, when manual testing might be simpler.

I'm not even sure I could meaningfully describe the tests manually, now that I think about it...

Reply:
@peterspicer-catalyst  hey! Sorry for the delay in response. I got what you mean here. And it makes sense. 
Anyway, would you be able then abstract just changed line in a function and cover it by some tests?